test/integration/volume/attach_detach_test.go

```diff
@@ -137,6 +142,98 @@ var defaultTimerConfig = attachdetach.TimerConfig{
+		ReconcilerMaxWaitForUnmountDuration:         6 * time.Second,
+		DesiredStateOfWorldPopulatorLoopSleepPeriod: 24 * time.Hour,
+		// Use high duration to disable DesiredStateOfWorldPopulator.findAndAddActivePods loop in test.
+		DesiredStateOfWorldPopulatorListPodsRetryDuration: 24 * time.Hour,
+	})
+
+	namespaceName := "test-node-oos"
+	nodeName := "node-sandbox"
+	node := &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: nodeName,
+			Annotations: map[string]string{
+				util.ControllerManagedAttachAnnotation: "true",
+			},
+		},
+	}
+	ns := framework.CreateNamespaceOrDie(testClient, namespaceName, t)
+	defer framework.DeleteNamespaceOrDie(testClient, ns, t)
+
+	_, err := testClient.CoreV1().Nodes().Create(context.TODO(), node, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to created node : %v", err)
+	}
+
+	pod := fakePodWithVol(namespaceName)
+	if _, err := testClient.CoreV1().Pods(pod.Namespace).Create(context.TODO(), pod, metav1.CreateOptions{}); err != nil {
+		t.Errorf("Failed to create pod : %v", err)
+	}
+
+	// start controller loop
+	podInformer := informers.Core().V1().Pods().Informer()
+	informers.Start(tCtx.Done())
+	informers.WaitForCacheSync(tCtx.Done())
+	go ctrl.Run(tCtx)
+	go pvCtrl.Run(tCtx)
+	go podgcCtrl.Run(tCtx)
+
+	waitToObservePods(t, podInformer, 1)
+	// wait for volume to be attached
+	waitForVolumeToBeAttached(tCtx, t, testClient, pod.Name, nodeName)
+
+	// Patch the node to mark the volume in use as attach-detach controller verifies if safe to detach the volume
+	// based on that.
+	node.Status.VolumesInUse = append(node.Status.VolumesInUse, "kubernetes.io/mock-provisioner/fake-mount")
+	node, err = testClient.CoreV1().Nodes().UpdateStatus(context.TODO(), node, metav1.UpdateOptions{})
+	if err != nil {
+		t.Fatalf("error in patch volumeInUse status to nodes: %s", err)
+	}
+	// Delete the pod with grace period time so that it is stuck in terminating state
+	gracePeriod := int64(300)
+	err = testClient.CoreV1().Pods(namespaceName).Delete(context.TODO(), pod.Name, metav1.DeleteOptions{
+		GracePeriodSeconds: &gracePeriod,
+	})
+	if err != nil {
+		t.Fatalf("error in deleting pod: %v", err)
+	}
+
+	// varify that DeletionTimestamp is not nil, means pod is in `Terminating` stsate
+	waitForPodDeletionTimestampToSet(tCtx, t, testClient, pod.Name, pod.Namespace)
+
+	// taint the node `out-of-service`
+	taint := v1.Taint{
+		Key:    v1.TaintNodeOutOfService,
+		Effect: v1.TaintEffectNoExecute,
+	}
+	node.Spec.Taints = append(node.Spec.Taints, taint)
+	if _, err := testClient.CoreV1().Nodes().Update(context.TODO(), node, metav1.UpdateOptions{}); err != nil {
+		t.Fatalf("error in patch oos taint to node: %v", err)
+	}
+	waitForNodeToBeTainted(tCtx, t, testClient, v1.TaintNodeOutOfService, nodeName)
+
+	// Verify is the pod was force deleted.
+	// When the node has out-of-service taint, and only if node is NotReady and pod is Terminating force delete will happen.
```

Comment:
There is something strange here that may need to be confirmed. The taint “node.kubernetes.io/out-of-service” effect should not take node status into account by https://github.com/kubernetes/kubernetes/blob/d87bf75c29cbc821ebf30b64a5d49a312e7b37e2/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L236, which is inconsistent with the pod gccontroller logic on https://github.com/kubernetes/kubernetes/blob/d87bf75c29cbc821ebf30b64a5d49a312e7b37e2/pkg/controller/podgc/gc_controller.go#L157

Reply:
But forceDetach depends on nodeIsHealth which checks ready? https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L162

Reply:
yes, what I mean here is that regardless of whether the node is ready or not, as long as the node is tainted with outofservice, it will go through the process of forcedetach, and won't be blocked by 
https://github.com/kubernetes/kubernetes/blob/ad9dad07a89acd183cf925a5726291de3798da2a/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L236
and 
https://github.com/kubernetes/kubernetes/blob/ad9dad07a89acd183cf925a5726291de3798da2a/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L271

Reply:
I see what you are saying. I think it should be fine though. The GC controller will be checking if node is not ready and if the taint is applied first. Then it will be forcefully deleting the pods. After that the AD reconciler will decide whether to delete the VolumeAttachment. It should just follow through and delete VolumeAttachment if the taint is applied.

Reply:
Thank you for your explanation, that's clear.